javaslang-benchmark/src/test/java/javaslang/JmhRunner.java

```diff
@@ -71,9 +72,28 @@ private static void runAndReport(Class<?> benchmarkClass, int warmupIterations,
         }
     }
 
-    public static <T> void assertEquals(T a, T b) {
-        if (!Objects.equals(a, b)) {
-            throw new IllegalStateException(a + " != " + b);
+    public static <T> void require(T value, Predicate<T> predicate) {
+        require(() -> predicate.test(value));
+    }
+
+    public static <T1, T2> void require(T1 value1, Predicate<T1> predicate1, T2 value2, Predicate<T2> predicate2) {
+        require(() -> predicate1.test(value1),
+                () -> predicate2.test(value2));
+    }
+
+    public static <T1, T2, T3> void require(T1 value1, Predicate<T1> predicate1, T2 value2, Predicate<T2> predicate2, T3 value3, Predicate<T3> predicate3) {
+        require(() -> predicate1.test(value1),
+                () -> predicate2.test(value2),
+                () -> predicate3.test(value3));
+    }
+
+    public static void require(BooleanSupplier... suppliers) {
```

Comment:
AFAIK the JVM recompiles in those cases. If it takes the same path multiple times, it can even delete unused (but valid) else statements (don't quote me on that :p).

```java
public static class DeadCode {
    @Benchmark
    public Object a() {
        final String s = new Exception().getStackTrace().toString();
        require(() -> !s.isEmpty());
        return s;
    }

    @Benchmark
    public Object b() {
        return new Exception().getStackTrace().toString();
    }

    public static void require(BooleanSupplier... suppliers) {
        if (WITH_ASSERTS) {
            for (int i = 0; i < suppliers.length; i++) {
                if (!suppliers[i].getAsBoolean()) {
                    throw new IllegalStateException("Failure in supplier #" + (i + 1) + "!");
                }
            }
        }
    }
}
```
gave the result:
```
Benchmark                    Mode  Cnt      Score      Error  Units
VectorBenchmark.DeadCode.a  thrpt   30  78716.918 ±  612.413  ops/s
VectorBenchmark.DeadCode.b  thrpt   30  78448.375 ± 2509.181  ops/s
```
i.e. they have the same speed.